include/GraphCoarsen.hpp

```diff
@@ -0,0 +1,208 @@
+
+    /** @brief Restrict a fine level mixed form vector to the coarse level
+        @param fine_vect fine level mixed form vector
+        @returns coarse_vect restricted mixed form vector
+    */
+    BlockVector Restrict(const BlockVector& fine_vect) const;
+
+    /** @brief Restrict a fine level mixed form vector to the coarse level
+        @param fine_vect fine level mixed form vector
+        @param coarse_vect restricted mixed form vector
+    */
+    void Restrict(const BlockVector& fine_vect, BlockVector& coarse_vect) const;
+
+    /** @brief Get the face to coarse dof relationship */
+    const SparseMatrix& GetFaceCDof() const { return face_cdof_; }
+
+    /** @brief Get Graph Topology */
+    const GraphTopology& GetGraphTopology() const { return gt_; }
+
+private:
+    template <class T>
+    using Vect2D = std::vector<std::vector<T>>;
+
+    void ComputeVertexTargets(const ParMatrix& M_ext, const ParMatrix& D_ext);
+    void ComputeEdgeTargets(const MixedMatrix& mgl,
+                            const ParMatrix& face_edge_perm);
+    void ScaleEdgeTargets(const SparseMatrix& D_local);
+
+
+    Vect2D<DenseMatrix> CollectSigma(const SparseMatrix& face_edge);
+    Vect2D<SparseMatrix> CollectD(const SparseMatrix& D_local);
+    Vect2D<std::vector<double>> CollectM(const SparseMatrix& M_local);
+
+    std::vector<double> Combine(const std::vector<std::vector<double>>& face_M,
+                                int num_face_edges) const;
+    SparseMatrix Combine(const std::vector<SparseMatrix>& face_D, int num_face_edges) const;
+
+    Vector MakeOneNegOne(int size, int split) const;
+
+    int GetSplit(int face) const;
+
+    void BuildAggBubbleDof();
+    void BuildFaceCoarseDof();
+    void BuildPvertex();
+    void BuildPedge(const MixedMatrix& mgl);
+
+    // These only depend on GraphTopology and are sent directly to
+    // the coarse mixed matrix
+    // {
+    SparseMatrix BuildAggCDofVertex() const;
+    SparseMatrix BuildAggCDofEdge() const;
+    ParMatrix BuildEdgeTrueEdge() const;
+    // }
+
+    SparseMatrix BuildCoarseD() const;
+    std::vector<DenseMatrix> BuildElemM(const MixedMatrix& mgl,
+                                        const SparseMatrix& agg_cdof_edge) const;
+
+    DenseMatrix RestrictLocal(const DenseMatrix& ext_mat,
+                              std::vector<int>& global_marker,
+                              const std::vector<int>& ext_indices,
+                              const std::vector<int>& local_indices) const;
+
+    std::vector<int> GetExtDofs(const ParMatrix& mat_ext, int row) const;
+
+    ParMatrix MakeExtPermutation(const ParMatrix& parmat) const;
+
+    GraphTopology gt_;
+
+    int max_evects_;
+    double spect_tol_;
+
+    SparseMatrix P_edge_;
+    SparseMatrix P_vertex_;
+    SparseMatrix face_cdof_;
+    SparseMatrix agg_bubble_dof_;
+
+    std::vector<DenseMatrix> vertex_targets_;
+    std::vector<DenseMatrix> edge_targets_;
+    std::vector<DenseMatrix> agg_ext_sigma_;
```

Comment:
These three were temporary objects before. Are you thinking about preparing for adaptive code? Otherwise at least their allocated memory can be destroyed after the coarsening process? 

I think most likely `agg_ext_sigma_` do not need to be kept even in adaptive algorithm?

Reply:
It is just easier to keep these as member functions right now to avoid passing them around to every function.  It may be useful to clear them after coarsening is complete.